spring-integration-kafka/src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundGateway.java

```diff
@@ -170,19 +181,24 @@ public void setBindSourceRecord(boolean bindSourceRecord) {
 		this.bindSourceRecord = bindSourceRecord;
 	}
 
-	@SuppressWarnings("deprecation")
 	@Override
 	protected void onInit() {
 		super.onInit();
-		MessageListener<K, V> kafkaListener = this.listener;
 		if (this.retryTemplate != null) {
-			kafkaListener =
-					new org.springframework.kafka.listener.adapter.RetryingMessageListenerAdapter<>(kafkaListener,
-							this.retryTemplate, this.recoveryCallback);
 			this.retryTemplate.registerListener(this.listener);
+			MessageChannel errorChannel = getErrorChannel();
+			if (this.recoveryCallback != null && errorChannel != null) {
+				this.recoveryCallback = new ErrorMessageSendingRecoverer(errorChannel, getErrorMessageStrategy());
+			}
 		}
 		ContainerProperties containerProperties = this.messageListenerContainer.getContainerProperties();
-		containerProperties.setMessageListener(kafkaListener);
+		Object existing = containerProperties.getMessageListener();
+		if (existing != null) {
+			logger.warn(() -> "Container's existing message listener ("
```

Comment:
Can't we throw an `IllegalArgumentException` instead to signal that provided listener is wrong?
I just try to avoid questions why their listener is out of use...